File: week1/hnnynh/prgs-단어변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

vector<bool> visited(51, false);
int answer = 50;

bool check(string s1, string s2){
    int cnt = 0;
    for(int i=0; i<s1.size(); i++){
        if(s1[i] != s2[i])
            cnt++;
    }
    
    if(cnt == 1)
        return true;
    return false;
}

void dfs(string begin, string target, vector<string> words, int count){
    if(begin == target){
        answer = count;
        return;
    }
    if(count >= answer)
        return;

    for(int i=0; i<words.size(); i++){
        if(!visited[i] && check(begin, words[i])){
            visited[i] = true;
            dfs(words[i], target, words, count+1);
            visited[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    
    if(find(words.begin(), words.end(), target) == words.end())
        return 0;
    
    dfs(begin, target, words, 0);
    
    return answer;
}

File: week1/gominzip/prgs-단어변환/prgs-단어변환.js
function solution(begin, target, words) {
  var answer;
  const wordLen = begin.length;

  if (!words.includes(target)) return 0;

  function dfs(path, words) {
    const latest = path[path.length - 1];
    if (latest === target) {
      if (!answer || answer > path.length) answer = path.length;
      return;
    }

    if (answer && path.length >= answer - 1) return;

    for (let word of words) {
      let differ = 0;
      for (let i = 0; i < wordLen; i++) {
        if (latest[i] != word[i]) differ++;
      }
      if (differ === 1) {
        dfs(
          [...path, word],
          words.filter((w) => w !== word)
        );
      }
    }

    return;
  }

  dfs([begin], words);
  return answer ? answer - 1 : 0;
}
